src/H5CXprivate.h

```diff
@@ -47,6 +50,235 @@ typedef struct H5CX_state_t {
 #endif                       /* H5_HAVE_PARALLEL */
 } H5CX_state_t;
 
+/* Typedef for context about each API call, as it proceeds */
+/* Fields in this struct are of several types:
+ * - The DXPL & LAPL ID are either library default ones (from the API context
+ *      initialization) or passed in from the application via an API call
+ *      parameter.  The corresponding H5P_genplist_t* is just the underlying
+ *      property list struct for the ID, to optimize retrieving properties
+ *      from the list multiple times.
+ *
+ * - Internal fields, used and set only within the library, for managing the
+ *      operation under way.  These do not correspond to properties in the
+ *      DXPL or LAPL and can have any name.
+ *
+ * - Cached fields, which are not returned to the application, for managing
+ *      the operation under way.  These correspond to properties in the DXPL
+ *      or LAPL, and are retrieved either from the (global) cache for a
+ *      default property list, or from the corresponding property in the
+ *      application's (non-default) property list.  Getting / setting these
+ *      properties within the library does _not_ affect the application's
+ *      property list.  Note that the naming of these fields, <foo> and
+ *      <foo>_valid, is important for the H5CX_RETRIEVE_PROP_VALID
+ *      macro to work properly.
+ *
+ * - "Return-only" properties that are returned to the application, mainly
+ *      for sending out "introspection" information ("Why did collective I/O
+ *      get broken for this operation?", "Which filters are set on the chunk I
+ *      just directly read in?", etc) Setting these fields will cause the
+ *      corresponding property in the property list to be set when the API
+ *      context is popped, when returning from the API routine.  Note that the
+ *      naming of these fields, <foo> and <foo>_set, is important for the
+ *      H5CX_TEST_SET_PROP and H5CX_SET_PROP macros to work properly.
+ *
+ * - "Return-and-read" properties that are returned to the application to send out introspection information,
+ *      but are also queried by the library internally. If the context value has been 'set' by an accessor,
+ *      all future queries will return the stored value from the context, to avoid later queries overwriting
+ *      that stored value with the value from the property list.
+ *
+ *      These properties have both a 'valid' and 'set' flag. <foo>_valid is true if the field has ever been
+ *      populated from its underlying property list. <foo>_set flag is true if this field has ever been set on
+ *      the context for application introspection. The naming of these fields is important for the
+ *      H5CX_RETRIEVE_PROP_VALID_SET macro to work properly.
+ *
+ *      If a field has been set on the context but never read internally, <foo>_valid will be false
+ *      despite the context containing a meaningful cached value.
+ */
+typedef struct H5CX_t {
```

Comment:
An interesting point I hadn't considered but may be a problem in the future - the size of this struct may soon grow large enough to cause stack warnings.

Reply:
I thought about that too, and I checked the size: ~600 bytes currently.   So, we should have a lot of margin for growth in the future.  If it gets more than 10x bigger, we'll have to think about it again.  :-)